### PR TITLE
Added main page module

### PR DESCRIPTION
**Pull request by maxonchil**
Added lazy loading
Added guard for login
Added jwt authentication
Added Login\register
Added pipes, interfaces
Created separate module for sharing material modules

### CODE REVIEW

File: src/app/components/login/login.component.html [hunk 1]
```diff
@@ -1,16 +1,32 @@
...
+    <div class="login__form" [formGroup]="loginForm">
         <mat-form-field class="login__field">
             <mat-label class="login__label">Login</mat-label>
-            <input class="login__input" placeholder="JojoBoy" matNativeControl type="text">
+            <input formControlName="username" class="login__input" placeholder="JojoBoy" matNativeControl type="text">
+            <mat-error *ngIf="
+                loginForm.controls.username.touched && loginForm.controls.username.errors?.required
+                ||
+                loginForm.controls.username.touched && loginForm.controls.username.errors?.minlength">
+                Min 3 characters required
```
Comment (MaxHouz) [inline src/app/components/login/login.component.html:10]: You should show one more error if the `required` validation did not pass.

File: src/app/components/login/login.component.html [hunk 2]
```diff
@@ -1,16 +1,32 @@
...
             <mat-label class="login__label">Password</mat-label>
-            <input class="login__input" placeholder="@NaGiBaToR228" matNativeControl type="password">
+            <input formControlName="password" class="login__input" placeholder="@NaGiBaToR228" matNativeControl
+                   type="password">
+            <mat-error *ngIf="
+                loginForm.controls.password.touched && loginForm.controls.password.errors?.required
+                ||
+                loginForm.controls.password.touched && loginForm.controls.password.errors?.minlength">
+                Min 6 characters required
+            </mat-error>
```
Comment (MaxHouz) [inline src/app/components/login/login.component.html:22]: The same, display different messages for both errors

File: src/app/components/login/login.component.ts [hunk 3]
```diff
@@ -1,15 +1,36 @@
...
+    selector: 'app-login',
+    templateUrl: './login.component.html',
+    styleUrls: ['./login.component.scss']
 })
 export class LoginComponent implements OnInit {
+    public loginForm: FormGroup;
+    public errorMessage: string;
 
-  constructor() { }
+    constructor(private fb: FormBuilder, private authService: AuthenticationService, private router: Router) {
```
Comment (MaxHouz) [inline src/app/components/login/login.component.ts:15]: it looks more neat if you format properties in this way:
````
constructor(
    private fb: FormBuilder, 
    private authService: AuthenticationService, 
    private router: Router
){}
````

File: src/app/components/login/login.component.ts [hunk 4]
```diff
@@ -1,15 +1,36 @@
...
+        this.loginForm = this.fb.group({
+            username: ['', [Validators.required, Validators.minLength(3)]],
+            password: ['', [Validators.required, Validators.minLength(6)]]
+        });
+    }
 
+    public loginUser(): void {
+        const username = this.loginForm.get('username').value;
+        const password = this.loginForm.get('password').value;
+        this.authService.loginUser(username, password).subscribe(({error, data}) => {
```
Comment (MaxHouz) [inline src/app/components/login/login.component.ts:28]: this is strange behaviour for API, it should not return error in success object, it should return it properly, using error code. And on UI there is `catchError` operator, or the callback as second observer argument to handle such error.

File: src/app/components/registration/registration.component.html [hunk 5]
```diff
@@ -20,28 +20,24 @@
...
-            </mat-form-field>
-        </div>
+        <mat-form-field class="register__field">
+            <mat-label class="register__label">Password</mat-label>
+            <input formControlName="password" class="register__input" placeholder="@NaGiBaToR228"
+                   matNativeControl
+                   type="password">
+            <mat-error *ngIf="regForm.controls.password.touched && regForm.controls.password.errors?.required
+            ||
+            regForm.controls.password.touched && regForm.controls.password.errors?.minlength
```
Comment (MaxHouz) [inline src/app/components/registration/registration.component.html:30]: Again, display messages for both errors

File: src/app/interfaces/auth-response.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+export interface AuthResponse {
+    message: string;
+    data: Token;
```
Comment (MaxHouz) [inline src/app/interfaces/auth-response.ts:3]: maybe you can just use type `string` as a token and not create a separate interface with only one field for it?

File: src/app/interfaces/todo-response.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { Todo } from './todo';
+
+export interface TodoResponse {
+    status: string;
+    message: string;
+    data?: Data;
+}
+
+interface Data {
```
Comment (MaxHouz) [inline src/app/interfaces/todo-response.ts:9]: `Data` is too generic name for the interface, always give interfaces as meaningful name, as you can

File: src/app/interfaces/todo.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+export interface Todo {
+    _id: string;
+    created_by: number;
+    title: string;
+    description: string;
+    completed: boolean;
+    date: string;
+    time: string;
```
Comment (MaxHouz) [inline src/app/interfaces/todo.ts:8]: A think you do not have the time separate from date. As for me much better is to store date in ISO string, which will include both time and date. And then you can just parse this string and get time from it.

File: src/app/interfaces/validate.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export interface Validate {
+    [key: string]: boolean;
```
Comment (MaxHouz) [inline src/app/interfaces/validate.ts:2]: Oh, you already know this hack 😉

File: src/app/main-page/components/add-todo/add-todo.component.html [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                    class="dialog__error"
+                    *ngIf="this.todoForm.controls.time.errors?.required"
+            >Current time required minimum
+            </mat-error
+            >
+        </mat-form-field>
+        <mat-form-field class="dialog__form-field">
+            <mat-label>Priority</mat-label>
+            <mat-select required formControlName="priority">
+                <mat-option [value]="priorities[0]">Light</mat-option>
```
Comment (MaxHouz) [inline src/app/main-page/components/add-todo/add-todo.component.html:60]: It is much better to store such kind of information in `enum`

File: src/app/main-page/components/add-todo/todo-priorities.enum.ts [hunk 11]
```diff
@@ -0,0 +1,6 @@
+export enum TodoPriorities {
+    Light,
+    Medium,
+    Important,
+    'Omae wa no shinderiu'
```
Comment (MaxHouz) [inline src/app/main-page/components/add-todo/todo-priorities.enum.ts:5]: Ahahah again this one, that I cannot understand))

File: src/app/main-page/components/header/header.component.html [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+            </div>
+            <div class="header__filter">
+                <mat-form-field>
+                    <mat-label>Filter note</mat-label>
+                    <mat-select
+                            [(ngModel)]="filterValue"
+                            (ngModelChange)="changeFilterQuery()"
+                    >
+                        <mat-option value="deadline_coming">Deadline coming</mat-option>
```
Comment (MaxHouz) [inline src/app/main-page/components/header/header.component.html:31]: move the values to a separate `enum`

File: src/app/main-page/components/header/header.component.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import {Observable} from 'rxjs';
+import {Todo} from '../../../interfaces/todo';
+
+@Component({
+    selector: 'app-header',
+    templateUrl: './header.component.html',
+    styleUrls: ['./header.component.scss']
+})
+export class HeaderComponent implements OnInit {
+    @Input() todos$: Observable<Todo[]>;
```
Comment (MaxHouz) [inline src/app/main-page/components/header/header.component.ts:14]: Do not pass observable here, this fieldshould be `@Input() todos: Todo[];` and all the async/observable stuff should be handled outside this component

File: src/app/main-page/components/main/main.component.html [hunk 14]
```diff
@@ -0,0 +1,7 @@
+<app-header (filterQueryChange)="filterQuery = $event" (searchQueryChange)="searchQuery = $event"
+            [todos$]="todos$"></app-header>
```
Comment (MaxHouz) [inline src/app/main-page/components/main/main.component.html:2]: just use `async` pipe here and pass only the value to component

File: src/app/main-page/components/todo-item/todo-item.component.html [hunk 15]
```diff
@@ -0,0 +1,9 @@
+<div class="todo">
+    <div class="todo__header">
+        {{todo.title}}
+        <buttton mat-icon-button>
+            <mat-icon>delete_forever</mat-icon>
```
Comment (MaxHouz) [inline src/app/main-page/components/todo-item/todo-item.component.html:5]: sounds not very optimistic 😂

File: src/app/main-page/main-page.module.ts [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+
+import {MainPageRoutingModule} from './main-page-routing.module';
+import {HeaderComponent} from './components/header/header.component';
+import {AddTodoComponent} from './components/add-todo/add-todo.component';
+import {MaterialSharedModule} from '../material-shared/material-shared.module';
+import {FormsModule, ReactiveFormsModule} from '@angular/forms';
+import {SearchPipe} from './pipes/search.pipe';
+import {FilterPipe} from './pipes/filter.pipe';
+import { TodoItemComponent } from './components/todo-item/todo-item.component';
```
Comment (MaxHouz) [inline src/app/main-page/main-page.module.ts:13]: Oops, this spaces should not be here)

File: src/app/main-page/pipes/filter.pipe.ts [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+@Pipe({
+    name: 'filter'
+})
+export class FilterPipe implements PipeTransform {
+
+
+    transform(todos: Todo[], filterValue: string): Todo[] {
+
+        switch (filterValue) {
+            case 'deadline_coming': {
```
Comment (MaxHouz) [inline src/app/main-page/pipes/filter.pipe.ts:14]: this one should be used from some common `enum`

File: src/app/material-shared/material-shared.module.ts [hunk 18]
```diff
@@ -3,19 +3,39 @@ import {CommonModule} from '@angular/common';
...
+import {MatNativeDateModule} from '@angular/material/core';
 
 
 @NgModule({
     declarations: [],
     imports: [
         CommonModule,
         MatFormFieldModule,
-        MatInputModule, MatButtonModule
+        MatInputModule,
```
Comment (MaxHouz) [inline src/app/material-shared/material-shared.module.ts:19]: move all of this modules to a constant on the top of the component like `const MATERIAL_MODULES = [...]`. And then just reuse it in `imports` and `exports`.
````
imports: [ ...MATERIAL_MODULES  ],
exports: [ ...MATERIAL_MODULES  ],
````
This way if you need to add a new module, you have to add it only in one place.

File: src/app/services/authorization.service.ts [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+    providedIn: 'root'
+})
+export class AuthorizationService {
+    private jwtHelper: JwtHelperService = new JwtHelperService();
+
+    constructor() {
+    }
+
+    isAuthorized(): boolean {
+        return !this.jwtHelper.isTokenExpired(localStorage.getItem('token'));
```
Comment (MaxHouz) [inline src/app/services/authorization.service.ts:15]: You should check on the server side if the token is expired. And then logout user, or show error message if he does some requests that need jwt

File: src/app/services/data.service.ts [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+})
+export class DataService {
+    private readonly jwtToken: string;
+
+    constructor(private http: HttpClient) {
+        this.jwtToken = localStorage.getItem('token');
+    }
+
+    public getTodosRequest(): Observable<TodoResponse> {
+        return this.http.get<TodoResponse>(`${env.apiURL}/todos/getAll`, {headers: {authentication: this.jwtToken}});
```
Comment (MaxHouz) [inline src/app/services/data.service.ts:19]: You can create an interceptor, that will add jwt header to all the requests and then then on server check for it if it is needed

File: src/app/components/login/login.component.html (hunk 1 above)
Comment (maxonchil) [reply src/app/components/login/login.component.html:10]: done

File: src/app/components/login/login.component.html (hunk 2 above)
Comment (maxonchil) [reply src/app/components/login/login.component.html:22]: done

File: src/app/components/login/login.component.ts (hunk 3 above)
Comment (maxonchil) [reply src/app/components/login/login.component.ts:15]: done

File: src/app/components/login/login.component.ts (hunk 4 above)
Comment (maxonchil) [reply src/app/components/login/login.component.ts:28]: Error was handled at server ad json object with error code coming as answer.

File: src/app/components/registration/registration.component.html (hunk 5 above)
Comment (maxonchil) [reply src/app/components/registration/registration.component.html:30]: done

File: src/app/interfaces/auth-response.ts (hunk 6 above)
Comment (maxonchil) [reply src/app/interfaces/auth-response.ts:3]: All responses with data  from server is coming at field "data" as object

File: src/app/interfaces/todo-response.ts (hunk 7 above)
Comment (maxonchil) [reply src/app/interfaces/todo-response.ts:9]: changed

File: src/app/interfaces/todo.ts (hunk 8 above)
Comment (maxonchil) [reply src/app/interfaces/todo.ts:8]: ok, good one, thx

File: src/app/main-page/components/add-todo/add-todo.component.html (hunk 10 above)
Comment (maxonchil) [reply src/app/main-page/components/add-todo/add-todo.component.html:60]: it's already stored in `enum`

File: src/app/main-page/components/add-todo/todo-priorities.enum.ts (hunk 11 above)
Comment (maxonchil) [reply src/app/main-page/components/add-todo/todo-priorities.enum.ts:5]: Here you go :)
https://www.youtube.com/watch?v=dNQs_Bef_V8

File: src/app/main-page/components/header/header.component.html (hunk 12 above)
Comment (maxonchil) [reply src/app/main-page/components/header/header.component.html:31]: done

File: src/app/main-page/components/header/header.component.ts (hunk 13 above)
Comment (maxonchil) [reply src/app/main-page/components/header/header.component.ts:14]: done

File: src/app/main-page/components/main/main.component.html (hunk 14 above)
Comment (maxonchil) [reply src/app/main-page/components/main/main.component.html:2]: done, thx

File: src/app/main-page/components/todo-item/todo-item.component.html (hunk 15 above)
Comment (maxonchil) [reply src/app/main-page/components/todo-item/todo-item.component.html:5]: I did my best)

File: src/app/main-page/main-page.module.ts (hunk 16 above)
Comment (maxonchil) [reply src/app/main-page/main-page.module.ts:13]: wops, my bad

File: src/app/main-page/pipes/filter.pipe.ts (hunk 17 above)
Comment (maxonchil) [reply src/app/main-page/pipes/filter.pipe.ts:14]: done

File: src/app/material-shared/material-shared.module.ts (hunk 18 above)
Comment (maxonchil) [reply src/app/material-shared/material-shared.module.ts:19]: goddamn that's looks so good!

File: src/app/services/data.service.ts (hunk 20 above)
Comment (maxonchil) [reply src/app/services/data.service.ts:19]: that so good, god...

File: src/app/services/authorization.service.ts (hunk 19 above)
Comment (maxonchil) [reply src/app/services/authorization.service.ts:15]: i don't know how to logout user, but i add a jwt exp check at server, and if it expired, user will not get any data from server

File: src/app/components/login/login.component.ts (hunk 4 above)
Comment (MaxHouz) [reply src/app/components/login/login.component.ts:28]: That’s still a little bit strange. As for me, an error should be an error and some kind of “error code” wrapped in success response. In fact I’ve already met such behavior in production projects. But in classical REST api, the error should be sent as error response.

**Pull request merged**